lib/rust/lexer/definition/src/lexer.rs

```diff
@@ -1157,3 +2320,416 @@ impl Default for BlockState {
+        self.starting_line_ending = line_ending;
+    }
+
+    /// Append a line of text to the current literal.
+    pub fn append_line(&mut self, line:Token) {
+        if self.has_empty_lines() {
+            let lines = self.take_empty_lines();
+            lines.into_iter().for_each(|l| self.append_line(l));
+        }
+        self.lines.push(line);
+    }
+
+    /// Append an empty line with `offset` leading spaces to the current text literal.
+    pub fn append_empty_line(&mut self, offset:usize, line_ending:token::LineEnding) {
+        let line = Token::BlankLine(offset,line_ending);
+        self.empty_lines.push(line);
+    }
+
+    /// Check if the current text state has unprocessed empty lines remaining.
+    pub fn has_empty_lines(&self) -> bool {
+        !self.empty_lines.is_empty()
+    }
+
+    /// Take the empty lines from the literal.
+    pub fn take_empty_lines(&mut self) -> Vec<Token> {
+        mem::take(&mut self.empty_lines)
+    }
+
+    /// Append a token to the current line of the text literal.
+    pub fn append_segment_to_line(&mut self, token:Token) {
+        if self.has_empty_lines() {
+            let lines = self.take_empty_lines();
+            lines.into_iter().for_each(|l| self.append_line(l));
+        }
+        self.segments.push(token);
+    }
+
+    /// Consume the last token from the currently active line.
+    pub fn consume_segment_from_line(&mut self) -> Option<Token> {
+        self.segments.pop()
+    }
+
+    /// Push a line ending onto the line ending stack.
+    pub fn record_line_ending(&mut self, line_ending:token::LineEnding) {
+        self.explicit_line_endings.push_back(line_ending);
+    }
+
+    /// Consume a line ending from the line ending stack.
+    pub fn consume_line_ending(&mut self) -> token::LineEnding {
+        self.explicit_line_endings.pop_front().unwrap_or(token::LineEnding::None)
+    }
+
+    /// Consume the current line of the text literal as a line.
+    pub fn consume_current_line(&mut self) -> Token {
+        let line_ending = self.consume_line_ending();
+        let tokens      = mem::take(&mut self.segments);
+        Token::Line(tokens,0,line_ending)
+    }
+
+    /// Submit the current line in the literal.
+    pub fn submit_current_line(&mut self) {
+        let line = self.consume_current_line();
+        self.append_line(line);
+    }
+
+    /// Set the style of text literal being lexed to `style`.
+    pub fn set_style(&mut self, style:token::TextStyle) {
+        self.style = Some(style);
+    }
+
+    /// Reset the text lexing state.
+    pub fn reset(&mut self) {
+        self.style                 = default();
+        self.offset                = default();
+        self.indent = default();
+        self.lines                 = default();
+        self.segments = default();
+        self.explicit_line_endings = default();
+    }
+}
+
+
+// === Trait Impls ===
+
+impl Into<Token> for TextState {
```

Comment:
never define `Into` impls. Always define `From` instead - they are more generic. See Rust docs for more info